### PR TITLE
Bedrock support

### PR DESCRIPTION
**Pull request by edward3h**
Support for current bedrock format, with a choice of resolutions.

(Also included vscode container setup since that's what I used to develop in)

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#-------------------------------------------------------------------------------------------------------------
```
Comment (jthistle) [inline .devcontainer/Dockerfile:1]: Just wondering about the whole `.devcontainer` directory, what's the point of it?

File: src/components/DownloadView/index.js [hunk 2]
```diff
@@ -103,6 +110,21 @@ export default ({ handleInput, onDownload, onClose }) => (
               />
             </td>
           </tr>
+          <tr>
```
Comment (jthistle) [inline src/components/DownloadView/index.js:113]: I would prefer it if this whole row were hidden when `enableResolution === false`, rather than just being disabled.

File: src/pages/Home/configs.js [hunk 3]
```diff
@@ -36,10 +36,29 @@ const MC_1_14_NAMES = {
...
+const BR_1_14_POSITIONS = {
+  '1x1': { size: _s(1, 1), positions: [_p(0, 0), _p(1, 0), _p(2, 0), _p(3, 0), _p(4, 0), _p(5, 0), _p(6, 0)] },
+  '1x2': { size: _s(1, 2), positions: [_p(0, 4), _p(1, 4)] },
+  '2x1': { size: _s(2, 1), positions: [_p(0, 2), _p(2, 2), _p(4, 2), _p(6, 2), _p(8, 2)] },
+  '2x2': { size: _s(2, 2), positions: [_p(0, 8), _p(2, 8), _p(4, 8), _p(6, 8), _p(8, 8), _p(10, 8)] },
+  '4x2': { size: _s(4, 2), positions: [_p(0, 6)] },
+  '4x3': { size: _s(4, 3), positions: [_p(12, 4), _p(12, 7)] },
+  '4x4': { size: _s(4, 4), positions: [_p(0, 12), _p(4, 12), _p(8, 12)] }
+};
+
```
Comment (jthistle) [inline src/pages/Home/configs.js:57]: :+1:

File: src/pages/Home/index.js [hunk 4]
```diff
@@ -68,6 +70,8 @@ function generateInitial() {
   return initial;
 }
 
+const DEFAULT_EXTENSION = 'zip';
```
Comment (jthistle) [inline src/pages/Home/index.js:73]: This should be a member of `DEFAULT_PACK_META` in `configs.js`.

File: src/pages/Home/index.js [hunk 5]
```diff
@@ -230,15 +235,72 @@ const Home = () => {
     });
   };
 
-  const createZip = async () => {
-    const packName = packMeta.name || DEFAULT_PACK_META.name;
-    const packDesc =
-      (packMeta.description ? packMeta.description + ' | ' : '') +
-      DEFAULT_PACK_META.description;
-    const packFormat = packMeta.pack_format || DEFAULT_PACK_META.pack_format;
+  const bedrockFileBuilder = async (root, packFormat, packDesc, packName, resolution) => {
```
Comment (jthistle) [inline src/pages/Home/index.js:238]: I notice these are independent of the `Home` component, so should probably be in their own file... but I'll do that myself later :)

I do like the idea of 'builder' functions, though.

File: src/pages/Home/index.js [hunk 6]
```diff
@@ -230,15 +235,72 @@ const Home = () => {
...
+    let imageString = canvas
+      .toDataURL()
+      .replace('data:image/png;base64,', '');
+    painting.file(`kz.png`, imageString, {
+      base64: true,
+    });
+
+    return userPaintingsCount;
+
+
```
Comment (jthistle) [inline src/pages/Home/index.js:300]: Unnecessary whitespace

File: src/pages/Home/index.js [hunk 7]
```diff
@@ -336,18 +413,32 @@ const Home = () => {
         break;
       case 'version':
         let pack_format = DEFAULT_PACK_META.pack_format;
+        newPackMeta.extension = DEFAULT_EXTENSION;
+        let showResolution = false;
         switch (event.value) {
           case '1_14':
             pack_format = 4;
+            newPackMeta.fileBuilder = javaFileBuilder;
```
Comment (jthistle) [inline src/pages/Home/index.js:421]: I'm a bit sceptical about this, since the `fileBuilder` isn't really part of the pack metadata. Same for `extension`. Maybe it would be better to read the `pack_format` at download-time to work out what `fileBuilder` and `extension` should be. You can just leave it if you want me to implement that instead.

File: src/pages/Home/index.js [hunk 8]
```diff
@@ -336,18 +413,32 @@ const Home = () => {
...
             break;
           default:
             console.error('Invalid pack version');
         }
         newPackMeta.pack_format = pack_format;
+        setShowResolutionSelect(showResolution);
         break;
+      case 'resolution':
+        newPackMeta.resolution = event.value;
+        break
```
Comment (jthistle) [inline src/pages/Home/index.js:441]: I thought prettier ran post-commit... evidently it hasn't quite worked. This line needs a semicolon.

File: src/pages/Home/index.js [hunk 9]
```diff
@@ -387,7 +479,7 @@ const Home = () => {
             onUpload={onImageUpload}
             text={
               selectedSize &&
-              uploadedImages[selectedSize.size][selectedSize.index]
+                uploadedImages[selectedSize.size][selectedSize.index]
```
Comment (jthistle) [inline src/pages/Home/index.js:482]: Not sure why the need for this change? Was it an automatic change made by prettier?

**Comment by jthistle** [comment]
Huh, having a recommitted everything myself locally, prettier runs fine and fixes it all up. It seems it didn't work for you... could you try squashing and committing everything again, and try to get prettier to run this time? Thank you!

**Comment by edward3h** [comment]
The .devcontainer is for
https://code.visualstudio.com/docs/remote/containers#_quick-start-open-a-git-repository-or-github-pr-in-an-isolated-container-volume

It's a pretty cool feature of vscode - you can check out a git branch in an isolated Docker container and do all your development in that so you have a consistent environment.

I would totally understand if you didn't want that committed in your project though.

File: src/components/DownloadView/index.js (hunk 2 above)
Comment (edward3h) [reply src/components/DownloadView/index.js:113]: Sure, that's easy enough to change.

File: src/pages/Home/index.js (hunk 4 above)
Comment (edward3h) [reply src/pages/Home/index.js:73]: Ok, will move.

File: src/pages/Home/index.js (hunk 5 above)
Comment (edward3h) [reply src/pages/Home/index.js:238]: Yeah they could do with modularising a bit better. If you have an idea how to do it I'll leave it with you. I thought about it a bit but couldn't make up my mind.

File: src/pages/Home/index.js (hunk 6 above)
Comment (edward3h) [reply src/pages/Home/index.js:300]: I'll clean it up.

File: src/pages/Home/index.js (hunk 9 above)
Comment (edward3h) [reply src/pages/Home/index.js:482]: I ran 'Format Document' in vscode, it may not be using the prettier config.

File: src/pages/Home/index.js (hunk 8 above)
Comment (edward3h) [reply src/pages/Home/index.js:441]: I have a guess that the Prettier config is not working right due to how I checked out in a restricted container (see other comment) - I'll see if I can figure that out.

File: src/pages/Home/index.js (hunk 9 above)
Comment (edward3h) [reply src/pages/Home/index.js:482]: I'll tidy up.

**Comment by edward3h** [comment]
I was motivated because my son's birthday is coming up, and due to lockdown his party is likely to be online in Minecraft. Wouldn't it be cool to have some custom birthday pictures :) All the kids his age are playing on iPads.

**Comment by jthistle** [comment]
That's a really cool idea! I'll try and get this merged as quickly as possible for you then.

**Comment by jthistle** [comment]
In regards to prettier, if you could please run `yarn run pretty-quick` once and then commit the changes, that should fix everything up fine. Normally that command runs on a post-commit hook, but evidently not when you're using it in a container.

In regards to the Docker container - no harm in leaving it in, it's nice and compartmentalized in its own directory :)

File: src/pages/Home/index.js (hunk 9 above)
Comment (jthistle) [reply src/pages/Home/index.js:482]: See below: `yarn run pretty-quick` will fix everything automatically for you.

**Comment by edward3h** [comment]
`yarn run pretty-quick` by itself was failing - looks like it was trying to compare against the master branch, but in my container there was no master branch because my branch was based on 'develop'.

Running `yarn run pretty-quick --branch develop` worked.

**Comment by jthistle** [comment]
Great, once you've addressed the comments I'll merge.

File: src/pages/Home/index.js (hunk 7 above)
Comment (edward3h) [reply src/pages/Home/index.js:437]: I'm leaving this as is.

File: src/pages/Home/index.js (hunk 5 above)
Comment (edward3h) [reply src/pages/Home/index.js]: I'm leaving this to you :)

**Comment by edward3h** [comment]
I've addressed the remaining comments.

Yes, I tested in bedrock edition BTW. Very funny seeing a photo of my cat on a wall.

**Comment by jthistle** [comment]
Merged in c79e64fd85b14713106155a1ead1cf709f538bad

**Comment by jthistle** [comment]
Oh yeah, and happy birthday to your son! Thank you again for your contribution.
